### PR TITLE
[Refactor] Rewrite abstraction layers in Meta Memcache

### PR DESCRIPTION
**Pull request by bisho**
## Motivation / Description
Instead of a huge class full of methods and responsibilities
we now have:
- ConnectionPoolProvider: Responsible for server selection
- Executors: Responsible of executing the commands on the appropriate
  pool. Some executors might implement extra routing logic, like
  the gutter pool that can handle errors and re-route requests to
  a different pool.
- Commands: Implement the actual commands, from the low-level meta
  commands to the high-level ones. These are implemented as mixins
  so you can easily add more custom commands as needed.
- CachePool is now mostly the command implementation with a executor
  (using composition) so you can instantate pools with different
  behavior depending on the executor. The idea is that people will
  implement their custom CachePools since is now very easy to pick
  different pieces, add mixins, etc.

### CODE REVIEW

File: src/meta_memcache/cache_pool.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from meta_memcache.routers.default import DefaultRouter
+from meta_memcache.routers.gutter import GutterRouter
+from meta_memcache.interfaces.cache_api import CacheApiProtocol
+from meta_memcache.protocol import Key
+from meta_memcache.serializer import BaseSerializer, MixedSerializer
+
+
+class CachePool(HighLevelCommandsMixin, MetaCommandsMixin, BaseCachePool):
+    @staticmethod
+    def cache_pool_from_servers(
```
Comment (tonycosentini) [inline src/meta_memcache/cache_pool.py:23]: I think there's still some value in having `sharded` or `hash` in the method name - some libraries don't default to a sharded pool (I remember python-binary-memcached defaults to replicating keys to all servers provided and not hashing).

**Review by tonycosentini — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "meta-memcache"
-version = "0.9.5"
+version = "1.0.0"
```
[the remark this answers is not in the document]
Comment (bisho) [reply pyproject.toml:3]: I break compatibility, so it needed to be major XD

File: src/meta_memcache/configuration.py [hunk 3]
```diff
@@ -145,9 +156,14 @@ class StalePolicy(NamedTuple):
...
 
-def default_binary_key_encoding(key: Key) -> bytes:
+def default_key_encoder(key: Key) -> Tuple[bytes, bool]:
     """
-    Generate binary key from Key, used for keys with unicode,
-    binary or those too long.
+    Generate valid memcache key as bytes for given Key.
     """
-    return hashlib.blake2b(key.key.encode(), digest_size=18).digest()
+    if key.is_unicode or len(key.key) > MAX_KEY_SIZE:
```
[the remark this answers is not in the document]
Comment (bisho) [reply src/meta_memcache/configuration.py:163]: Yeah, there is room for improvement there. We can definitely improve the key handling and support binary as well.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bisho) [reply pyproject.toml:3]: Perhaps just 0.10? Not sure if people will assume 0.X are compatible... :)

**Pull request merged**